arch-vega: Improve SDWA, SDWAB, and DPP

abmerop opened the pull request:
This PR has four components:

- Implement a helper method for SDWAB similar to SDWA helpers. SDWAB is used for VOPC instructions only (vector compares).
- Update two instructions commonly using SDWAB to use helper (v_cmp_ne_u16 and v_cmp_eq_u16).
- Add panics to *all* VOP1 and VOP2 instructions which do not implement SDWA or DPP if they use an SDWA or DPP register.
- Add panics to *all* VOPC instructions which do not implement SDWAB or DPP if they are an SDWA or DPP register.

Only VOP1, VOP2, and VOPC may use SDWA, SDWAB, or DPP. The panics should therefore cover all instructions which have missing implementations for these modes. The intent is to exit gem5 instead of continuing simulation will data that is likely incorrect. Continuing simulation only makes debugging gem5 more difficult.

abmerop committed:
arch-vega: Panic on SDWAB / DPP VOPC unimplemented

If SDWAB or DPP are used on a VOPC instruction and those are not
implemented, it is highly likely to be a problem for the application.
Rather than continue to execute and cause undefined behavior, exit the
simulation with a panic showing the line of the instruction causing the
issue.

Change-Id: Ib3f94df7445d068b26907470c1f733be16cd2fc2

abmerop committed:
arch-vega: Implement SDWAB helper

Implement a SDWAB helper which accepts a dynamic instruction and a
lambda function defining a comparison function taking two values and
returning a comparison result of 0 or 1 for false or true.

Current instructions which implement SDWA do so on a per-instruction
basis which adds a lot of redundant code. This allows for generic SDWAB
implementations for VOPC instructions.

All modifiers are implemented assuming that SDWBA VOPC instruction
comparison types may be U32, I32, F32, U16, I16, F16 (which exist) but
is extendible to I8, U8, or F8.

Change-Id: Idab58a327c29dd19a1a5457237f3799a04f2031b

abmerop committed:
arch-vega: Add SDWAB for v_cmp_{eq,ne}_u16

This shows an example of how to use the previous commit which adds an
SDWAB helper. The execute() method of both are the same with the
exception of the lambda function passed to the helper method.

Change-Id: I5ffe361440b4020b9f7669c0ed946aa6b3bbec25

mattsinc commented on src/arch/amdgpu/vega/insts/op_encodings.hh:
```diff
@@ -421,6 +422,145 @@ namespace VegaISA
+            int low_bit = 0, high_bit = 0;
+            uint32_t rv = dword;
+
+            if (sel < SDWA_WORD_0) {
```
Ideally we should add comments explaining what each of these are doing.

abmerop replied:
Done

mattsinc commented on src/arch/amdgpu/vega/insts/op_encodings.hh:
```diff
@@ -421,6 +422,145 @@ namespace VegaISA
+
+            rv = bits(dword, high_bit, low_bit);
+
+            // Apply ABS, then NEG, then SEXT.
```
looks like this comment should go before line 455?

abmerop committed:
arch-vega: Panic unimplemented SDWA/DPP for VOP1/VOP2

Add a panic if SDWA or DPP is used for an instruction which does not
implement support for it. If an application uses SDWA or DPP it likely
does not operate in the same way as the base instruction and therefore
gem5 should panic rather than continue. It is likely data is incorrect
which will make it more difficult to debug an application.

Change-Id: I68ac448b0d62941761ef4efa0169f95796270f48

mattsinc commented on src/arch/amdgpu/vega/insts/op_encodings.hh:
```diff
@@ -421,6 +422,145 @@ namespace VegaISA
+                    }
+                } else {
+                    // It is not entirely clear what to do here. Literal
+                    // extensions for FP operands appends zeros to mantissa
```
appends --> append

mattsinc commented on src/arch/amdgpu/vega/insts/vopc.cc:
```diff
@@ -3874,17 +4089,27 @@ namespace VegaISA
         Wavefront *wf = gpuDynInst->wavefront();
-        ConstVecOperandU16 src0(gpuDynInst, instData.SRC0);
-        ConstVecOperandU16 src1(gpuDynInst, instData.VSRC1);
+        ConstVecOperandU32 src0(gpuDynInst, instData.SRC0);
```
why is U32 correct for a U16 instruction?  Maybe because SDWA needs an extra bit internally?

abmerop replied:
GPU always reads 32-bits into a register.  In order to be able to select upper or lower 16-bits, we need to read the full value which the U16 types will not do.  This will become uint16_t when the lambda function is called (upper 16-bits get cropped).

abmerop replied:
Actually, it seems this doesn't matter here.  The `sdwabHelper` will read the 32-bit value if it is SDWAB. Otherwise it reads (correctly) the lower 16-bits.  

I noticed I did not change this in the other compare so I can either revert this change here or apply the change to the other instruction making it U32.  Leaning towards the former.

mattsinc replied:
I think reverting is fine then.

abmerop replied:
Done
